sqlglot/parser.py

```diff
@@ -6094,3 +6104,65 @@ def _parse_truncate_table(self) -> t.Optional[exp.TruncateTable] | exp.Expressio
             option=option,
             partition=partition,
         )
+
+    def _parse_exclude_element(self) -> exp.ExcludeElement:
+        this = self._parse_expression() if self._match(TokenType.L_PAREN) else self._parse_column()
+
+        opclass = None
+        if not self._match_texts(self.OPCLASS_FOLLOW_KEYWORDS, advance=False) and not self._match(
+            TokenType.WITH, advance=False
+        ):
+            opclass = self._parse_var(any_token=True)
+
+        order = self._prev.text if self._match_set((TokenType.ASC, TokenType.DESC)) else None
+
+        nulls = (
+            self._prev.text
+            if self._match_text_seq("NULLS", "FIRST") or self._match_text_seq("NULLS", "LAST")
+            else None
+        )
+
+        self._match(TokenType.WITH)
+
+        operator = self._parse_var(any_token=True)
+
+        return self.expression(
+            exp.ExcludeElement,
+            this=this,
+            opclass=opclass,
+            order=order,
+            nulls=nulls,
+            operator=operator,
+        )
+
+    def _parse_exclude(self) -> exp.ExcludeConstraint:
+        using = self._parse_var(any_token=True) if self._match(TokenType.USING) else None
+
+        self._match_l_paren()
+        exclude_elems = self._parse_csv(self._parse_exclude_element)
+        self._match_r_paren()
+
```

Comment:
Ditto about not duplicating this section - we should factor out what we have in `_parse_index` and reuse it.